director/jobs/models.py

```diff
@@ -0,0 +1,433 @@
+    CANCELLED = "CANCELLED"
+    # Job was revoked.
+    REVOKED = "REVOKED"
+    # Job was terminated (custom).
+    TERMINATED = "TERMINATED"
+
+    # Job was rejected.
+    REJECTED = "REJECTED"
+    # Job is waiting for retry.
+    RETRY = "RETRY"
+
+    @classmethod
+    def has_ended(cls, status: str) -> bool:
+        return status in [
+            member.value
+            for member in (
+                cls.SUCCESS,
+                cls.FAILURE,
+                cls.CANCELLED,
+                cls.REVOKED,
+                cls.TERMINATED,
+            )
+        ]
+
+
+class Job(models.Model):
+    """
+    A job, usually, but not necessarily associated with a project.
+
+    Because some jobs may not be associated with a project, the
+    `key` field is a token that provides access to the job outside
+    of the usual project permissions.
+
+    If a job is created here in Django, the `creator` field should be
+    populated with the current user. Jobs created as part of a pipline
+    may not have a creator.
+
+    The fields `method`, `params`, `result` correspond to
+    the same properties in [JSON RPC](https://www.jsonrpc.org/specification).
+
+    The `log` field is an array of JSON objects, each corresponding to
+    a log entry. Each entry is exected to have a structure of a
+    [Logga](https://github.com/stencila/logga) log entry. It will include
+    any errors while running the job.
+
+    The fields `queue`, `worker` and `retries` provide for additional
+    information on where and how the job was executed.
+    """
+
+    project = models.ForeignKey(
+        Project,
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE,
+        related_name="jobs",
+        help_text="The project this job is associated with.",
+    )
+
+    creator = models.ForeignKey(
+        User,
+        null=True,  # Should only be null if the creator is deleted
+        blank=True,
+        on_delete=models.SET_NULL,
+        related_name="jobs_created",
+        help_text="The user who created the job.",
+    )
+
+    created = models.DateTimeField(
+        null=True, auto_now_add=True, help_text="The time the job was created."
+    )
+
+    zone = models.ForeignKey(
+        Zone,
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL,
+        related_name="jobs",
+        help_text="The project this job is associated with.",
+    )
+
+    queue = models.CharField(
+        max_length=64,
+        blank=True,
+        null=True,
+        help_text="The identifier of the queue the job was posted to.",
+    )
+
+    began = models.DateTimeField(null=True, help_text="The time the job began.")
+    ended = models.DateTimeField(null=True, help_text="The time the job ended.")
+    status = models.CharField(
+        max_length=32,
+        choices=JobStatus.as_choices(),
+        blank=True,
+        null=True,
+        help_text="The current status of the job.",
+    )
+
+    method = models.CharField(
```

Comment:
See comment above, a `max_length` of 32 specified for safety even though none of the `JobMethod` members are that long. Using `StringField` you don't need to worry about it.